File: docs/current_docs/ci/quickstart/snippets/daggerize/php/src/HelloDagger.php
<?php

declare(strict_types=1);

namespace DaggerModule;

use Dagger\Attribute\DaggerFunction;
use Dagger\Attribute\DaggerObject;
use Dagger\Attribute\Doc;
use Dagger\Container;
use Dagger\Directory;

use function Dagger\dag;

#[DaggerObject]
class HelloDagger
{
    #[DaggerFunction]
    #[Doc('Publish the application container after building and testing it on-the-fly')]
    public function publish(Directory $source): string
    {
        $this->test($source);

        return $this
            ->build($source)
            ->publish('ttl.sh/hello-dagger-' . rand(0, 10000000));
    }

    #[DaggerFunction]
    #[Doc('Build the application container')]
    public function build(Directory $source): Container
    {
        $build = $this
            ->buildEnv($source)
            ->withExec(['npm', 'run', 'build'])
            ->directory('./dist');

        return dag()
            ->container()
            ->from('nginx:1.25-alpine')
            ->withDirectory('/usr/share/nginx/html', $build)
            ->withExposedPort(80);
    }

    #[DaggerFunction]
    #[Doc('Return the result of running unit tests')]
    public function test(Directory $source): string
    {
        return $this
            ->buildEnv($source)
            ->withExec(['npm', 'run', 'test:unit', 'run'])
            ->stdout();
    }

    #[DaggerFunction]
    #[Doc('Build a ready-to-use development environment')]
    public function buildEnv(Directory $source): Container
    {
        $nodeCache = dag()
            ->cacheVolume('node');

        return dag()
            ->container()
            ->from('node:21-slim')
            ->withDirectory('/src', $source)
            ->withMountedCache('/root/.npm', $nodeCache)
            ->withWorkdir('/src')
            ->withExec(['npm', 'install']);
    }
}

File: docs/current_docs/quickstart/ci/snippets/java/src/main/java/io/dagger/modules/hellodagger/HelloDagger.java
package io.dagger.modules.hellodagger;

import static io.dagger.client.Dagger.dag;

import io.dagger.client.Container;
import io.dagger.client.DaggerQueryException;
import io.dagger.client.Directory;
import io.dagger.client.CacheVolume;
import io.dagger.module.annotation.DefaultPath;
import io.dagger.module.annotation.Function;
import io.dagger.module.annotation.Object;
import java.util.List;
import java.util.concurrent.ExecutionException;

/** HelloDagger main object */
@Object
public class HelloDagger {
  /** Publish the application container after building and testing it on-the-fly */
  @Function
  public String publish(@DefaultPath("/") Directory source)
      throws InterruptedException, ExecutionException, DaggerQueryException {
    this.test(source);
    return this.build(source).
        publish("ttl.sh/hello-dagger-%d".formatted((int) (Math.random() * 10000000)));
  }

  /** Build the application container */
  @Function
  public Container build(@DefaultPath("/") Directory source)
      throws InterruptedException, ExecutionException, DaggerQueryException {
    Directory build = this
        .buildEnv(source)
        .withExec(List.of("npm", "run", "build"))
        .directory("./dist");
    return dag().container()
        .from("nginx:1.25-alpine")
        .withDirectory("/usr/share/nginx/html", build)
        .withExposedPort(80);
  }

  /** Return the result of running unit tests */
  @Function
  public String test(@DefaultPath("/") Directory source)
      throws InterruptedException, ExecutionException, DaggerQueryException {
    return this
        .buildEnv(source)
        .withExec(List.of("npm", "run", "test:unit", "run"))
        .stdout();
  }

  /** Build a ready-to-use development environment */
  @Function
  public Container buildEnv(@DefaultPath("/") Directory source)
      throws InterruptedException, ExecutionException, DaggerQueryException {
    CacheVolume nodeCache = dag().cacheVolume("node");
    return dag().container()
        .from("node:21-slim")
        .withDirectory("/src", source)
        .withMountedCache("/root/.npm", nodeCache)
        .withWorkdir("/src")
        .withExec(List.of("npm", "install"));
  }
}
